src/components/tiles/table/use-content-change-handlers.ts

```diff
@@ -124,22 +125,36 @@ export const useContentChangeHandlers = ({
     if (!readOnly && consumerTile) {
       const sharedModelManager = consumerTile.tileEnv?.sharedModelManager;
       if (sharedModelManager?.isReady) {
+        // If the consumer tile does not support multiple shared data sets, remove it from
+        // any existing shared data set before linking.
+        if (!getTileContentInfo(consumerTile.type)?.supportsMultipleDataSets) {
+          const existingSharedTable = sharedModelManager?.findFirstSharedModelByType(SharedDataSet, tileInfo.id);
+          if (existingSharedTable) {
+            sharedModelManager?.removeTileSharedModel(consumerTile, existingSharedTable);
+          }
+        }
         const sharedTable = sharedModelManager?.findFirstSharedModelByType(SharedDataSet, modelRef.current.id);
         sharedTable && sharedModelManager?.addTileSharedModel(consumerTile, sharedTable);
       }
     }
   }, [getContent, readOnly, modelRef]);
 
   const unlinkTile = useCallback((tileInfo: ITileLinkMetadata) => {
-    const consumerTile = getTileContentById(getContent(), tileInfo.id);
-    if (!readOnly && consumerTile) {
-      const sharedModelManager = consumerTile.tileEnv?.sharedModelManager;
+    const linkedTile = getTileContentById(getContent(), tileInfo.id);
+    if (!readOnly && linkedTile) {
+      const sharedModelManager = linkedTile.tileEnv?.sharedModelManager;
       if (sharedModelManager?.isReady) {
         const sharedTable = sharedModelManager?.findFirstSharedModelByType(SharedDataSet, modelRef.current.id);
-        sharedTable && sharedModelManager?.removeTileSharedModel(consumerTile, sharedTable);
+        // If providerId matches model.id, we're the data provider and should remove the other tile
+        // from the shared data set. Otherwise, we're the consumer and should remove ourselves.
+        if (sharedTable && sharedTable.providerId === model.id) {
+          sharedModelManager?.removeTileSharedModel(linkedTile, sharedTable);
+        } else if (sharedTable) {
+          sharedModelManager?.removeTileSharedModel(model.content, sharedTable);
```

Comment:
```suggestion
        if (sharedTable && sharedTable.providerId === modelRef.current.id) {
          sharedModelManager?.removeTileSharedModel(linkedTile, sharedTable);
        } else if (sharedTable) {
          sharedModelManager?.removeTileSharedModel(modelRef.current.content, sharedTable);
```
Then you wouldn't have to add the additional dependencies to the `useEffect()`. If you don't like the redundancy, you could add something like this at the top of the effect:
```typescript
    const { id: modelId, content: modelContent } = modelRef.current
```